modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py

```diff
@@ -42,6 +54,82 @@ class PandasOnRayDataframePartitionManager(GenericRayDataframePartitionManager):
     _execution_wrapper = RayWrapper
     materialize_futures = RayWrapper.materialize
 
+    if LazyExecution.get() in ("On", "Axis"):
+        # Try calculating the indices without triggering the lazy split.
+
+        @classmethod
+        @_inherit_docstrings(GenericRayDataframePartitionManager.get_indices)
+        def get_indices(cls, axis, partitions, index_func=None):
```

Comment:
Have you tried making lazy changes to the already existing `get_indices`? (without overriding)

Reply:
What this function is trying to do is to avoid the partitions concatenation. It could be possible in the case when all the partitions are the result of a deferred split operation. Look at the description of the `find_non_split_block()` function. There is an example of such an execution tree. If we can find in the tree the non-split partition, we can just get the index out of there and, thus, avoid the concatenation.